tests/test_proxy_use.py

```diff
@@ -0,0 +1,366 @@
+
+  @classmethod
+  def setUpClass(cls):
+    """
+    Setup performed before the first test function (TestWithProxies class
+    method) runs.
+    Launch http, https, and proxy servers in the current working directory.
+    We'll set up four servers:
+     - HTTP server (simple_server.py)
+     - HTTPS server (simple_https_server.py)
+     - HTTP proxy server (proxy_server.py)
+         (that supports HTTP CONNECT to funnel HTTPS connections)
+     - HTTPS proxy server (proxy_server.py)
+         (trusted by the client to intercept and resign connections)
+    """
+
+    unittest_toolbox.Modified_TestCase.setUpClass()
+
+    # Launch a simple HTTP server (serves files in the current dir).
+    cls.http_port = random.randint(30000, 45000)
+    cls.http_server_proc = subprocess.Popen(
+        ['python', 'simple_server.py', str(cls.http_port)],
+        stderr=subprocess.PIPE)
+
+    # Launch an HTTPS server (serves files in the current dir).
+    cls.https_port = cls.http_port + 1
+    cls.https_server_proc = subprocess.Popen(
+        ['python', 'simple_https_server.py', str(cls.https_port)],
+        stderr=subprocess.PIPE)
+
+
+    # Launch an HTTP proxy server derived from inaz2/proxy2.
+    # This one is able to handle HTTP CONNECT requests, and so can pass HTTPS
+    # requests on to the target server.
+    cls.http_proxy_port = cls.http_port + 2
+    cls.http_proxy_proc = subprocess.Popen(
+        ['python', 'proxy_server.py', str(cls.http_proxy_port)],
+        stderr=subprocess.PIPE)
+    # Note that the http proxy server's address uses http://, regardless of the
+    # type of connection used with the target server.
+    cls.http_proxy_addr = 'http://127.0.0.1:' + str(cls.http_proxy_port)
+
+
+    # Launch an HTTPS proxy server, also derived from inaz2/proxy2.
+    # (An HTTPS proxy performs its own TLS connection with the client and must
+    # be trusted by it, and is capable of tampering.)
+    # We instruct the proxy server to expect certain certificates from the
+    # target server.
+    # 1st arg: port
+    # 2nd arg: whether to intercept (HTTPS proxy) or relay (TCP tunnel using
+    #   HTTP CONNECT verb, to facilitate an HTTPS connection between the client
+    #   and server which the proxy cannot inspect)
+    # 3rd arg: (optional) certificate file for telling the proxy what target
+    #   server certs to accept in its HTTPS connection to the target server.
+    #   This is only relevant if the proxy is in intercept mode.
+    cls.https_proxy_port = cls.http_port + 3
+    cls.https_proxy_proc = subprocess.Popen(
+        ['python', 'proxy_server.py', str(cls.https_proxy_port), 'intercept',
+        os.path.join('ssl_certs', 'ssl_cert.crt')],
+        stderr=subprocess.PIPE)
+    # Note that the https proxy server's address uses https://, regardless of
+    # the type of connection used with the target server.
+    cls.https_proxy_addr = 'https://127.0.0.1:' + str(cls.https_proxy_port)
```

Comment:
This uses "127.0.0.1", i.e. one of the IPv4 addresses of `localhost`. Why was IPv4 chosen here over the symbolic name? Will the code work with IPv6 too?

Reply:
I don't think the included test proxy will work in IPv6 as configured either way, no.

Reply:
The proxy tests run an IPv4 proxy and I think will fail with IPv6. It's probably true that the symbolic name should be used. I don't recall why I made this 127.0.0.1, but switching to localhost (both server and client side) seems to break things for some reason.